### PR TITLE
PTCD-739 Add Notifications to Provider Dashboard-Beta

### PR DESCRIPTION
**Pull request by jameslawesynergy**
**Please note that this is an attempt at a faithful reproduction (without change in behaviour) of the existing notification logic implemented in the dashboard component:**

https://github.com/SkillsFundingAgency/dfc-coursedirectory/blob/v3.8.0/src/Dfc.CourseDirectory.Web/ViewComponents/Dashboard/Dashboard.cs

https://github.com/SkillsFundingAgency/dfc-coursedirectory/blob/v3.8.0/src/Dfc.CourseDirectory.Web/ViewComponents/Dashboard/Default.cshtml

Updated QANotiffications with new copy + added component to dashboard-beta
Added migration notification to dashboard-beta
Added past course start date notification to dashboard-beta
Added Courses BulkUpload notifications to dashboard-beta
Added ApprenticeshipsBulkUpload notifications to dashboard-beta
Set up mock for IBinaryDataProvider + updated dashboard-beta tests

### CODE REVIEW

File: src/Dfc.CourseDirectory.Core/BinaryStorageProvider/IBinaryStorageProvider.cs [hunk 1]
```diff
@@ -7,5 +8,6 @@ public interface IBinaryStorageProvider
     {
         Task<bool> TryDownloadFile(string path, Stream destination);
         Task UploadFile(string path, Stream source);
+        Task<IEnumerable<BlobFileInfo>> ListFiles(string path);
```
Comment (gunndabad) [inline src/Dfc.CourseDirectory.Core/BinaryStorageProvider/IBinaryStorageProvider.cs:11]:
```suggestion
        Task<IReadOnlyList<BlobFileInfo>> ListFiles(string path);
```

File: src/Dfc.CourseDirectory.Core/DataStore/Sql/Queries/GetProviderDashboardCounts.cs [hunk 2]
```diff
@@ -1,9 +1,13 @@
 ﻿using System;
+using System.Collections.Generic;
+using Dfc.CourseDirectory.Core.Models;
 
 namespace Dfc.CourseDirectory.Core.DataStore.Sql.Queries
 {
-    public class GetProviderDashboardCounts : ISqlQuery<(int CourseRunCount, int ApprenticeshipCount, int VenueCount)>
+    public class GetProviderDashboardCounts : ISqlQuery<(IReadOnlyDictionary<CourseStatus, int> CourseRunCounts, IReadOnlyDictionary<ApprenticeshipStatus, int> ApprenticeshipCounts, int VenueCount, int PastStartDateCourseRunCount, int BulkUploadCoursesErrorCount, int BulkUploadCourseRunsErrorCount, int ApprenticeshipsBulkUploadErrorCount)>
```
Comment (gunndabad) [inline src/Dfc.CourseDirectory.Core/DataStore/Sql/Queries/GetProviderDashboardCounts.cs:7]: Given the size of this tuple, maybe a `DashboardCounts` class would be better here?

File: src/Dfc.CourseDirectory.Core/DataStore/Sql/QueryHandlers/GetProviderDashboardCountsHandler.cs [hunk 3]
```diff
@@ -1,39 +1,75 @@
...
+            var sql = @$"
 DECLARE @ProviderUkprn INT
 
-SELECT TOP 1 @ProviderUkprn = Ukprn FROM Pttcd.Providers WHERE ProviderId = @ProviderId
+SELECT TOP 1 @ProviderUkprn = Ukprn FROM Pttcd.Providers WHERE ProviderId = @{nameof(query.ProviderId)}
 
-SELECT COUNT(*) FROM Pttcd.Courses c
-JOIN Pttcd.CourseRuns cr ON c.CourseId = cr.CourseId
-WHERE c.ProviderUkprn = @ProviderUkprn AND cr.CourseRunStatus = 1
+SELECT		cr.CourseRunStatus, COUNT(*) as Count
```
Comment (gunndabad) [inline src/Dfc.CourseDirectory.Core/DataStore/Sql/QueryHandlers/GetProviderDashboardCountsHandler.cs:23]: Nit: indentation is off here

File: src/Dfc.CourseDirectory.Core/DataStore/Sql/QueryHandlers/GetProviderDashboardCountsHandler.cs [hunk 4]
```diff
@@ -1,39 +1,75 @@
...
+FROM		Pttcd.CourseRuns cr
+INNER JOIN	Pttcd.Courses c ON c.CourseId = cr.CourseId
+WHERE		c.ProviderUkprn = @ProviderUkprn
+GROUP BY	cr.CourseRunStatus
 
-SELECT COUNT(*) FROM Pttcd.Apprenticeships a
-WHERE a.ProviderUkprn = @ProviderUkprn AND a.ApprenticeshipStatus & 1 <> 0
+SELECT		a.ApprenticeshipStatus, COUNT(*) as Count
```
Comment (gunndabad) [inline src/Dfc.CourseDirectory.Core/DataStore/Sql/QueryHandlers/GetProviderDashboardCountsHandler.cs:29]: Nit: indentation is off here

File: src/Dfc.CourseDirectory.Core/DataStore/Sql/QueryHandlers/GetProviderDashboardCountsHandler.cs [hunk 5]
```diff
@@ -1,39 +1,75 @@
...
+GROUP BY	a.ApprenticeshipStatus
 
-SELECT COUNT(*) FROM Pttcd.Venues v
-WHERE v.ProviderUkprn = @ProviderUkprn AND v.VenueStatus = 1";
+SELECT      COUNT(*)
+FROM        Pttcd.Venues v
+WHERE       v.ProviderUkprn = @ProviderUkprn
+AND         v.VenueStatus = 1
+
+SELECT		COUNT(*)
```
Comment (gunndabad) [inline src/Dfc.CourseDirectory.Core/DataStore/Sql/QueryHandlers/GetProviderDashboardCountsHandler.cs:39]: Nit: indentation is off here

File: src/Dfc.CourseDirectory.Core/DataStore/Sql/QueryHandlers/GetProviderDashboardCountsHandler.cs [hunk 6]
```diff
@@ -1,39 +1,75 @@
...
+FROM		Pttcd.CourseRuns cr
+INNER JOIN	Pttcd.Courses c ON c.CourseId = cr.CourseId
+WHERE		c.ProviderUkprn = @ProviderUkprn
+AND			c.CourseStatus = c.CourseStatus & {(int)CourseStatus.Live}
```
Comment (gunndabad) [inline src/Dfc.CourseDirectory.Core/DataStore/Sql/QueryHandlers/GetProviderDashboardCountsHandler.cs:43]: Nit: indentation is off here

File: src/Dfc.CourseDirectory.WebV2/Features/NewApprenticeshipProvider/QANotifications.cshtml [hunk 7]
```diff
@@ -1,53 +1,45 @@
...
     @if (Model.Status.HasFlag(ApprenticeshipQAStatus.Submitted) || Model.Status.HasFlag(ApprenticeshipQAStatus.InProgress))
     {
         <pttcd-notification title="Quality assurance in progress" id="pttcd-new-apprenticeship-provider__qa-notifications-submitted">
-            <p>Once your quality assurance is complete you will be contacted with the outcome.</p>
+            <p class="govuk-body govuk-!-margin-bottom-0">Once your quality assurance is complete you will be contacted with the outcome.</p>
         </pttcd-notification>
     }
     else if (Model.Status.HasFlag(ApprenticeshipQAStatus.Failed))
     {
-        <pttcd-notification title="Quality assurance outcome" id="pttcd-new-apprenticeship-provider__qa-notifications-failed">
```
Comment (gunndabad) [inline src/Dfc.CourseDirectory.WebV2/Features/NewApprenticeshipProvider/QANotifications.cshtml:17]: Given that this component is already in use, presumably there's no issue with us shipping this updated copy for the old dashboard too?

File: src/Dfc.CourseDirectory.WebV2/Features/ProviderDashboard/Dashboard.cshtml [hunk 8]
```diff
@@ -1,5 +1,69 @@
...
+@if (Model.BulkUploadInProgress)
+{
+    <pttcd-notification title="Processing bulk upload of courses">
+        <div class="govuk-warning-text">
+            <span class="govuk-warning-text__icon" aria-hidden="true">!</span>
+            <strong class="govuk-warning-text__text">
+                <span class="govuk-warning-text__assistive">Warning</span>
+                Do not add a new course or attempt another bulk upload during this process
+            </strong>
+        </div>
```
Comment (gunndabad) [inline src/Dfc.CourseDirectory.WebV2/Features/ProviderDashboard/Dashboard.cshtml:30]:
```suggestion
        <govuk-warning-text icon-fallback-text="Warning">
            Do not add a new course or attempt another bulk upload during this process
        </govuk-warning-text>
```

File: src/Dfc.CourseDirectory.WebV2/Features/ProviderDashboard/Dashboard.cshtml [hunk 9]
```diff
@@ -1,5 +1,69 @@
...
+        <pttcd-notification title="Successful bulk upload of courses">
+            <p class="govuk-body">@Model.BulkUploadReadyToGoLiveCourseRunCount courses have been uploaded and are ready to publish to the course directory.</p>
+            <p class="govuk-body govuk-!-margin-bottom-0"><a asp-action="Index" asp-controller="PublishCourses" class="govuk-link">Publish your courses</a></p>
+        </pttcd-notification>
+    }
+}
+
+@if (Model.ApprenticeshipsBulkUploadErrorCount > 0)
+{
+<pttcd-notification title="Error on bulk upload of apprenticeships">
```
Comment (gunndabad) [inline src/Dfc.CourseDirectory.WebV2/Features/ProviderDashboard/Dashboard.cshtml:54]: Nit: indentation

**Review by gunndabad — APPROVED** [review]
Assuming tests are forthcoming then 👍

File: src/Dfc.CourseDirectory.Core/BinaryStorageProvider/IBinaryStorageProvider.cs (hunk 1 above)
Comment (jameslawesynergy) [reply src/Dfc.CourseDirectory.Core/BinaryStorageProvider/IBinaryStorageProvider.cs:11]: Any particular reason?

File: src/Dfc.CourseDirectory.Core/DataStore/Sql/Queries/GetProviderDashboardCounts.cs (hunk 2 above)
Comment (jameslawesynergy) [reply src/Dfc.CourseDirectory.Core/DataStore/Sql/Queries/GetProviderDashboardCounts.cs:7]: I was thinking that myself!

File: src/Dfc.CourseDirectory.Core/BinaryStorageProvider/IBinaryStorageProvider.cs (hunk 1 above)
Comment (gunndabad) [reply src/Dfc.CourseDirectory.Core/BinaryStorageProvider/IBinaryStorageProvider.cs:11]: Means `Count` is more efficient and prevents having to protect against double enumeration

File: src/Dfc.CourseDirectory.Core/DataStore/Sql/QueryHandlers/GetProviderDashboardCountsHandler.cs (hunk 3 above)
Comment (jameslawesynergy) [reply src/Dfc.CourseDirectory.Core/DataStore/Sql/QueryHandlers/GetProviderDashboardCountsHandler.cs:22]: Note to self, beware copy and past from SSMS!

File: src/Dfc.CourseDirectory.WebV2/Features/ProviderDashboard/Dashboard.cshtml (hunk 8 above)
Comment (jameslawesynergy) [reply src/Dfc.CourseDirectory.WebV2/Features/ProviderDashboard/Dashboard.cshtml:30]: Nice.

File: src/Dfc.CourseDirectory.WebV2/Features/ProviderDashboard/Dashboard.cshtml (hunk 9 above)
Comment (jameslawesynergy) [reply src/Dfc.CourseDirectory.WebV2/Features/ProviderDashboard/Dashboard.cshtml:54]: dsjhfsdfsdkulf!! Visual Studio!

File: src/Dfc.CourseDirectory.Core/BinaryStorageProvider/IBinaryStorageProvider.cs (hunk 1 above)
Comment (jameslawesynergy) [reply src/Dfc.CourseDirectory.Core/BinaryStorageProvider/IBinaryStorageProvider.cs:11]: > Means Count is more efficient

`Count` doesn't exist on `IEnumerable<T>`, `Count()` does but would require an enumeration
> prevents having to protect against double enumeration

I like this argument 👍

**Review by gunndabad — APPROVED** [review]

**Review by gunndabad — APPROVED** [review]

**Pull request merged**